### DISCUSSION TITLE
Point Pick " pl.enable_point_picking" accuracy issues.

### DISCUSSION THREAD

**Discussion by ndahiya3** [pyvista/pyvista#2825] [Q&A]
I am trying to build an application for selecting landmark points on a surface (potentially multiple surfaces in the same view). But the accuracy of picked points seems to be way off. The actual point shows up super far away compared to where I click. There are clearly a bunch of points close to the mouse location but none of them seem to be picked. The picked point is literally on the other side of the surface which is not even visible. I have a couple of screenshots to illustrate. This is example the problem is really with actual use case with multiple surface. The screenshots are with pyvista plotter, but in my actual application, I am using pyvistaQT, QtInteractor, but the problem seems to be in both cases.

I see that there is a 'tolerance' parameter in the 'enable_point_picking' function but it doesn't seem to really help.

This is the place where I am trying to click:
![Point_pick_click_position](https://user-images.githubusercontent.com/15024623/174496427-a8fdd1d6-4453-4292-b466-b4a55551d057.png)

This is the place where the point shows up (other side of the surface). In other cases, it's much severe. The point is nowhere close to click location.
![Point_Pick_Actual_Position](https://user-images.githubusercontent.com/15024623/174496432-03de3903-8e8a-4ab6-b2f3-cae13ad7dd3b.png)

**Comment by ndahiya3** [comment]
Update: plotter.enable_cell_picking(through=False, callback=self.cell_pick_callback, show_message=False, style='wireframe'), in contrast to point picking seems way more robust. It seems to pick the correct triangles under the mouse.

Triangle get's picked and displayed at the exact mouse location:
![Picked_Position_Triangle](https://user-images.githubusercontent.com/15024623/174497196-378baab9-8eb0-4d87-ada3-9be3a71033ab.png)

**Comment by ndahiya3** [comment]
For the time being, I have rigged the cell_picking functionality to capture mouse position as well and then find the closes point from the triangle cell points. But this is not ideal situation.

**Comment by akaszynski** [comment]
Have you tried surface mesh picking?
https://docs.pyvista.org/examples/02-plot/surface-picking.html

**Reply by ndahiya3** [reply]
I tried surface picking as well. It seems more stable than point picking. But it picks points away from the surface (close to the edges). I think there's tolerance parameter that could help with but I haven't been able to get it right. It also, this picks points anywhere on the surface. For this application, I need to pick points that exist on the mesh.
![enable_surface_picking](https://user-images.githubusercontent.com/15024623/175783936-64cce764-60af-4890-a0af-005a4534c422.png)

 For now, with enable cell picking, I track the mouse location and select the cell point closest to the mouse location. Cell picking doesn't allow left clicking though. Some users want left clicking rather than keyboard shortcut.

For the point picking example, I'll try pure VTK to see if the problem is coming from there.

**Reply by akaszynski** [reply]
> Cell picking doesn't allow left clicking though. Some users want left clicking rather than keyboard shortcut.

That's a feature to add for sure.

If you can get it to work in pure VTK, please let us know and post the code here (doesn't have to be clean), so we can then implement it in this repo.

Thanks for posting what works with cell picking!
